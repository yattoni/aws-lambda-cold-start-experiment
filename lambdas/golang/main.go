package main

import (
	"fmt"
	"os"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/sns"
)

func handle(sqsEvent events.SQSEvent) error {
	for _, message := range sqsEvent.Records {
		fmt.Printf("The message %s for event source %s = %s \n", message.MessageId, message.EventSource, message.Body)
	}
	sess := session.Must(session.NewSession())
	snsClient := sns.New(sess)

	message := "Hello from golang lambda!"
	topicArn := os.Getenv("TARGET_TOPIC")

	fmt.Printf("Sending message to %s\n", topicArn)

	publishOutput, _ := snsClient.Publish(&sns.PublishInput{
		Message:  &message,
		TopicArn: &topicArn,
	})

	fmt.Printf("Sent message with ID %s\n", *publishOutput.MessageId)
	return nil
}

func main() {
	lambda.Start(handle)
}
