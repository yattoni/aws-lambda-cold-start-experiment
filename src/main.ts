import { env } from 'process';
import { App, Construct, Duration, Stack, StackProps } from 'monocdk';
import { Code, Function, Runtime } from 'monocdk/aws-lambda';
import { SqsEventSource } from 'monocdk/lib/aws-lambda-event-sources';
import { GoFunction } from 'monocdk/lib/aws-lambda-go';
import { PythonFunction } from 'monocdk/lib/aws-lambda-python';
import { Topic } from 'monocdk/lib/aws-sns';
import { SqsSubscription } from 'monocdk/lib/aws-sns-subscriptions';
import { Queue } from 'monocdk/lib/aws-sqs';

export class AwsLambdaColdStartExperiment extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    const goTopic = new Topic(this, 'GoTopic');
    const goQueue = new Queue(this, 'GoQueue');
    goTopic.addSubscription(new SqsSubscription(goQueue));

    const pythonTopic = new Topic(this, 'PythonTopic');
    const pythonQueue = new Queue(this, 'PythonQueue');
    pythonTopic.addSubscription(new SqsSubscription(pythonQueue));

    const javaTopic = new Topic(this, 'JavaTopic');
    const javaQueue = new Queue(this, 'JavaQueue');
    javaTopic.addSubscription(new SqsSubscription(javaQueue));

    const goFunction = new GoFunction(this, 'GoFunction', {
      entry: 'lambdas/golang',
      environment: {
        TARGET_TOPIC: pythonTopic.topicArn,
      },
    });
    pythonTopic.grantPublish(goFunction);

    const pythonFunction = new PythonFunction(this, 'PythonFunction', {
      entry: 'lambdas/python',
      environment: {
        TARGET_TOPIC: javaTopic.topicArn,
      },
    });
    javaTopic.grantPublish(pythonFunction);

    const javaFunction = new Function(this, 'JavaFunction', {
      runtime: Runtime.JAVA_11,
      // code: Code.fromAsset('lambdas/java/lib/build/libs/lib-0.1.0-all.jar'),
      code: Code.fromAsset('lambdas/java', {
        bundling: {
          command: [
            '/bin/sh', '-c', [
              './gradlew clean shadowJar --no-daemon',
              'cp /asset-input/lib/build/libs/lib-0.1.0-all.jar /asset-output/',
              'ls /asset-output/',
            ].join(' && '),
          ],
          image: Runtime.JAVA_11.bundlingImage,
          volumes: [
            {
              hostPath: `/Users/${env.USER}/.gradle/`,
              containerPath: '/root/.gradle',
            },
          ],
          user: 'root',
        },
      }),
      handler: 'com.yattoni.awslambdacoldstartexperiment.LambdaHandler',
      environment: {
        TARGET_TOPIC: goTopic.topicArn,
      },
      timeout: Duration.seconds(10),
      memorySize: 1024,
    });
    goTopic.grantPublish(javaFunction);

    goFunction.addEventSource(new SqsEventSource(goQueue, {
      batchSize: 10,
      maxBatchingWindow: Duration.seconds(10),
    }));
    pythonFunction.addEventSource(new SqsEventSource(pythonQueue, {
      batchSize: 10,
      // maxBatchingWindow: Duration.seconds(10),
    }));
    javaFunction.addEventSource(new SqsEventSource(javaQueue, {
      batchSize: 10,
      maxBatchingWindow: Duration.seconds(10),
    }));
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new AwsLambdaColdStartExperiment(app, 'aws-lambda-cold-start-experiment', { env: devEnv });
// new MyStack(app, 'my-stack-prod', { env: prodEnv });

app.synth();