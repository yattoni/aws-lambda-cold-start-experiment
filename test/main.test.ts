import '@aws-cdk/assert/jest';
import { App } from 'monocdk';
import { AwsLambdaColdStartExperiment } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new AwsLambdaColdStartExperiment(app, 'test');

  expect(stack).not.toHaveResource('AWS::S3::Bucket');
  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});