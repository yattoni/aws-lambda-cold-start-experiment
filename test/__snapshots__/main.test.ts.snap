// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters155a8fc577cb053208d378291d4fab6cea5c16c0d8020567af3aef910eaaa8bbArtifactHash625A6FC7": Object {
      "Description": "Artifact hash for asset \\"155a8fc577cb053208d378291d4fab6cea5c16c0d8020567af3aef910eaaa8bb\\"",
      "Type": "String",
    },
    "AssetParameters155a8fc577cb053208d378291d4fab6cea5c16c0d8020567af3aef910eaaa8bbS3Bucket619F6663": Object {
      "Description": "S3 bucket for asset \\"155a8fc577cb053208d378291d4fab6cea5c16c0d8020567af3aef910eaaa8bb\\"",
      "Type": "String",
    },
    "AssetParameters155a8fc577cb053208d378291d4fab6cea5c16c0d8020567af3aef910eaaa8bbS3VersionKeyA68FC406": Object {
      "Description": "S3 key for asset version \\"155a8fc577cb053208d378291d4fab6cea5c16c0d8020567af3aef910eaaa8bb\\"",
      "Type": "String",
    },
    "AssetParameters2b9841bb375c4a0e16a89ae1f8291e5a27c89ece004d53c8a9450a99c8701062ArtifactHash04C38409": Object {
      "Description": "Artifact hash for asset \\"2b9841bb375c4a0e16a89ae1f8291e5a27c89ece004d53c8a9450a99c8701062\\"",
      "Type": "String",
    },
    "AssetParameters2b9841bb375c4a0e16a89ae1f8291e5a27c89ece004d53c8a9450a99c8701062S3Bucket84DB482A": Object {
      "Description": "S3 bucket for asset \\"2b9841bb375c4a0e16a89ae1f8291e5a27c89ece004d53c8a9450a99c8701062\\"",
      "Type": "String",
    },
    "AssetParameters2b9841bb375c4a0e16a89ae1f8291e5a27c89ece004d53c8a9450a99c8701062S3VersionKeyCB1097D8": Object {
      "Description": "S3 key for asset version \\"2b9841bb375c4a0e16a89ae1f8291e5a27c89ece004d53c8a9450a99c8701062\\"",
      "Type": "String",
    },
    "AssetParameters9be4f8c7ee96ae181ccc61ee2eece9f46ac405d8b050ab88a51b7bd978ce3997ArtifactHashF32599F1": Object {
      "Description": "Artifact hash for asset \\"9be4f8c7ee96ae181ccc61ee2eece9f46ac405d8b050ab88a51b7bd978ce3997\\"",
      "Type": "String",
    },
    "AssetParameters9be4f8c7ee96ae181ccc61ee2eece9f46ac405d8b050ab88a51b7bd978ce3997S3Bucket12352654": Object {
      "Description": "S3 bucket for asset \\"9be4f8c7ee96ae181ccc61ee2eece9f46ac405d8b050ab88a51b7bd978ce3997\\"",
      "Type": "String",
    },
    "AssetParameters9be4f8c7ee96ae181ccc61ee2eece9f46ac405d8b050ab88a51b7bd978ce3997S3VersionKey5BEDB411": Object {
      "Description": "S3 key for asset version \\"9be4f8c7ee96ae181ccc61ee2eece9f46ac405d8b050ab88a51b7bd978ce3997\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "GoFunctionCA95FBAA": Object {
      "DependsOn": Array [
        "GoFunctionServiceRoleDefaultPolicyB6328196",
        "GoFunctionServiceRole9256C2F0",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters2b9841bb375c4a0e16a89ae1f8291e5a27c89ece004d53c8a9450a99c8701062S3Bucket84DB482A",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters2b9841bb375c4a0e16a89ae1f8291e5a27c89ece004d53c8a9450a99c8701062S3VersionKeyCB1097D8",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters2b9841bb375c4a0e16a89ae1f8291e5a27c89ece004d53c8a9450a99c8701062S3VersionKeyCB1097D8",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "TARGET_TOPIC": Object {
              "Ref": "PythonTopic459A6A89",
            },
          },
        },
        "Handler": "bootstrap",
        "Role": Object {
          "Fn::GetAtt": Array [
            "GoFunctionServiceRole9256C2F0",
            "Arn",
          ],
        },
        "Runtime": "provided.al2",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GoFunctionServiceRole9256C2F0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GoFunctionServiceRoleDefaultPolicyB6328196": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "PythonTopic459A6A89",
              },
            },
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "GoQueueC4DF8C4A",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GoFunctionServiceRoleDefaultPolicyB6328196",
        "Roles": Array [
          Object {
            "Ref": "GoFunctionServiceRole9256C2F0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GoFunctionSqsEventSourcetestGoQueue3050D643F90013E2": Object {
      "Properties": Object {
        "BatchSize": 10,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "GoQueueC4DF8C4A",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "GoFunctionCA95FBAA",
        },
        "MaximumBatchingWindowInSeconds": 10,
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "GoQueueC4DF8C4A": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "GoQueuePolicy58DF93F2": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sqs:SendMessage",
              "Condition": Object {
                "ArnEquals": Object {
                  "aws:SourceArn": Object {
                    "Ref": "GoTopic7C4768E1",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "sns.amazonaws.com",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "GoQueueC4DF8C4A",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": Array [
          Object {
            "Ref": "GoQueueC4DF8C4A",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "GoQueuetestGoTopic607F805C8C15193D": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "GoQueueC4DF8C4A",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "TopicArn": Object {
          "Ref": "GoTopic7C4768E1",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "GoTopic7C4768E1": Object {
      "Type": "AWS::SNS::Topic",
    },
    "JavaFunction11388F4F": Object {
      "DependsOn": Array [
        "JavaFunctionServiceRoleDefaultPolicyEE6BDC15",
        "JavaFunctionServiceRole0DBB0C5A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters9be4f8c7ee96ae181ccc61ee2eece9f46ac405d8b050ab88a51b7bd978ce3997S3Bucket12352654",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9be4f8c7ee96ae181ccc61ee2eece9f46ac405d8b050ab88a51b7bd978ce3997S3VersionKey5BEDB411",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9be4f8c7ee96ae181ccc61ee2eece9f46ac405d8b050ab88a51b7bd978ce3997S3VersionKey5BEDB411",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "TARGET_TOPIC": Object {
              "Ref": "GoTopic7C4768E1",
            },
          },
        },
        "Handler": "com.yattoni.awslambdacoldstartexperiment.LambdaHandler",
        "MemorySize": 1024,
        "Role": Object {
          "Fn::GetAtt": Array [
            "JavaFunctionServiceRole0DBB0C5A",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "JavaFunctionServiceRole0DBB0C5A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "JavaFunctionServiceRoleDefaultPolicyEE6BDC15": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "GoTopic7C4768E1",
              },
            },
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "JavaQueue1A8AFB0E",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JavaFunctionServiceRoleDefaultPolicyEE6BDC15",
        "Roles": Array [
          Object {
            "Ref": "JavaFunctionServiceRole0DBB0C5A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JavaFunctionSqsEventSourcetestJavaQueue62C96A15A444D78C": Object {
      "Properties": Object {
        "BatchSize": 10,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "JavaQueue1A8AFB0E",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "JavaFunction11388F4F",
        },
        "MaximumBatchingWindowInSeconds": 10,
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "JavaQueue1A8AFB0E": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "JavaQueuePolicyE5558966": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sqs:SendMessage",
              "Condition": Object {
                "ArnEquals": Object {
                  "aws:SourceArn": Object {
                    "Ref": "JavaTopic4B3D1781",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "sns.amazonaws.com",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "JavaQueue1A8AFB0E",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": Array [
          Object {
            "Ref": "JavaQueue1A8AFB0E",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "JavaQueuetestJavaTopic641E282EEFA91D8E": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "JavaQueue1A8AFB0E",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "TopicArn": Object {
          "Ref": "JavaTopic4B3D1781",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "JavaTopic4B3D1781": Object {
      "Type": "AWS::SNS::Topic",
    },
    "PythonFunction0BCF77FD": Object {
      "DependsOn": Array [
        "PythonFunctionServiceRoleDefaultPolicy0C392FE9",
        "PythonFunctionServiceRole4A77BE2A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters155a8fc577cb053208d378291d4fab6cea5c16c0d8020567af3aef910eaaa8bbS3Bucket619F6663",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters155a8fc577cb053208d378291d4fab6cea5c16c0d8020567af3aef910eaaa8bbS3VersionKeyA68FC406",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters155a8fc577cb053208d378291d4fab6cea5c16c0d8020567af3aef910eaaa8bbS3VersionKeyA68FC406",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "TARGET_TOPIC": Object {
              "Ref": "JavaTopic4B3D1781",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "PythonFunctionServiceRole4A77BE2A",
            "Arn",
          ],
        },
        "Runtime": "python3.7",
      },
      "Type": "AWS::Lambda::Function",
    },
    "PythonFunctionServiceRole4A77BE2A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PythonFunctionServiceRoleDefaultPolicy0C392FE9": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "JavaTopic4B3D1781",
              },
            },
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PythonQueueA573BBF0",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PythonFunctionServiceRoleDefaultPolicy0C392FE9",
        "Roles": Array [
          Object {
            "Ref": "PythonFunctionServiceRole4A77BE2A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PythonFunctionSqsEventSourcetestPythonQueueD891ECD03A93A370": Object {
      "Properties": Object {
        "BatchSize": 10,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "PythonQueueA573BBF0",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "PythonFunction0BCF77FD",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "PythonQueueA573BBF0": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "PythonQueuePolicy5D7BC756": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sqs:SendMessage",
              "Condition": Object {
                "ArnEquals": Object {
                  "aws:SourceArn": Object {
                    "Ref": "PythonTopic459A6A89",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "sns.amazonaws.com",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PythonQueueA573BBF0",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": Array [
          Object {
            "Ref": "PythonQueueA573BBF0",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "PythonQueuetestPythonTopic09C91C55C228C87D": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "PythonQueueA573BBF0",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "TopicArn": Object {
          "Ref": "PythonTopic459A6A89",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "PythonTopic459A6A89": Object {
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;
